### PR TITLE
Fix s3 requests without region

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category (leave one):
- Improvement

### Changelog entry:
Use `GetObjectAttributes` request instead of `HeadObject` request to get the size of an object in AWS S3.
This change fixes handling endpoints without explicit region, for example

```
select * from s3('https://test.s3.amazonaws.com/mydata.csv')
```

### CODE REVIEW

File: src/IO/S3Common.cpp [hunk 1]
```diff
@@ -894,61 +919,84 @@ namespace S3
...
         throw S3Exception(error.GetErrorType(),
             "Failed to check existence of key {} in bucket {}: {}",
             key, bucket, error.GetMessage());
     }
+
+    std::pair<bool /* exists */, Aws::S3::S3Error> checkObjectExists(const Aws::S3::S3Client & client, const String & bucket, const String & key, const String & version_id, bool for_disk_s3)
+    {
+        auto outcome = getObjectAttributes(client, bucket, key, version_id, for_disk_s3);
+        if (outcome.IsSuccess())
+            return {true, {}};
```
Comment (antonio2368) [inline src/IO/S3Common.cpp:964]: Maybe return `{}` for error if it's NotFound?

Comment (vitlibar) [reply src/IO/S3Common.cpp:964]: This function is called when we're quite sure that an object must exist. So we need to show an error if it's not.

File: src/IO/S3Common.cpp [hunk 2]
```diff
@@ -894,61 +919,84 @@ namespace S3
...
-        auto outcome = headObject(client, bucket, key, version_id, for_disk_s3);
+        auto [exists, error] = checkObjectExists(client, bucket, key, version_id, for_disk_s3);
 
-        if (outcome.IsSuccess())
+        if (exists)
             return true;
 
-        const auto & error = outcome.GetError();
-        if (isNotFoundError(error.GetErrorType()))
+        if (!throw_on_error || isNotFoundError(error.GetErrorType()))
```
[the remark this answers is not in the document]
Comment (vitlibar) [reply src/IO/S3Common.cpp:952]: I'll remove `throw_on_error` parameter here, we don't really need it anyway.

File: src/IO/S3Common.h [hunk 3]
```diff
@@ -121,21 +121,35 @@ struct URI
...
+
+size_t getObjectSize(const Aws::S3::S3Client & client, const String & bucket, const String & key, const String & version_id = "", bool for_disk_s3 = false, bool throw_on_error = true);
 
-Aws::S3::Model::HeadObjectOutcome headObject(const Aws::S3::S3Client & client, const String & bucket, const String & key, const String & version_id = "", bool for_disk_s3 = false);
+bool objectExists(const Aws::S3::S3Client & client, const String & bucket, const String & key, const String & version_id = "", bool for_disk_s3 = false);
 
-S3::ObjectInfo getObjectInfo(const Aws::S3::S3Client & client, const String & bucket, const String & key, const String & version_id, bool throw_on_error, bool for_disk_s3);
+/// Throws an exception if a specified object doesn't exist. `description` is used as a part of the error message.
+void checkObjectExists(const Aws::S3::S3Client & client, const String & bucket, const String & key, const String & version_id = "", bool for_disk_s3 = false, const std::string_view & description = {});
```
Comment (antonio2368) [inline src/IO/S3Common.h:147]:
```suggestion
void checkObjectExists(const Aws::S3::S3Client & client, const String & bucket, const String & key, const String & version_id = "", bool for_disk_s3 = false, std::string_view description = {});
```

File: src/IO/S3Common.cpp (hunk 1 above)
Comment (vitlibar) [reply src/IO/S3Common.cpp]: I changed the code so now the function just throws an exception if the object doesn't exist without returning anything.

**Review by antonio2368 — APPROVED** [review]

File: src/IO/S3Common.h (hunk 3 above)
Comment (vitlibar) [reply src/IO/S3Common.h]: ok

**Comment by vitlibar** [comment]
This is bad. It appears Minio doesn't support `GetObjectAttributes` requests, that's why many tests failed.

**Pull request merged**